### PR TITLE
MODINVOSTO-48 - Schema Changes for Batch Voucher Support

### PR DESCRIPTION
**Pull request by craigmcnally**
## Purpose
In order to support batch vouchers, several new schemas are needed.  There are also a couple changes to existing invoice/voucher schemas.  See [MODINVOSTO-48](https://issues.folio.org/browse/MODINVOSTO-48) 

See the [wiki](https://wiki.folio.org/display/FOLIJET/Batch+Voucher+Exports) for additional information.

## Approach
* Update existing schemas and exmaples
* Create new schemas and examples
* For batchGroupId, use the static UUID of the "system" record as a default.  A data migration script will need (in the related mod-invoice-storage PR) to address this for pre-existing data that doesn't already have batchGroupId specified.
* the common/uuid.json schema cannot be used for batchGroupId since we want to specify a default and that doesn't seem to work in conjunction with $ref.  Instead, the pattern from the uuid schema was used directly in the invoice and voucher schemas for this field.

## TODOS and Open Questions
- [ ] Will need to be merged before related PRs covering other parts of MODINVOSTO-48 can be merged (e.g. in folio-org/mod-invoice-storage)

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [X] Examples exist for all schemas
  - [X] Descriptions exist for all schema properties
  - [X] All schemas pass raml-cop linting
- Does this introduce breaking changes?
  - [ ] Were there any schema changes?
  - [X] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by pautau-s** [comment]
It is necessary to update the schemes in the mod-finance, because the default value for a new field will be set by storage-module, if the mod-finance does not know about this field we will get an error.

File: mod-invoice-storage/schemas/batch_group_collection.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "description": "collection of batch groups",
+  "type": "object",
+  "properties": {
+    "vouchers": {
```
[the remark this answers is not in the document]
Comment (craigmcnally) [reply mod-invoice-storage/schemas/batch_group_collection.json]: Good catch... CPM

**Comment by craigmcnally** [comment]
> It is necessary to update the schemes in the mod-finance, because the default value for a new field will be set by storage-module, if the mod-finance does not know about this field we will get an error.

That's a good point.  I'll create a PR for this.

File: mod-invoice-storage/schemas/batched_voucher.json [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      "description": "The name of the vendor",
+      "type": "string"
+    },
+    "voucherDate": {
+      "description": "The date the invoice status changes to approved",
+      "type": "string",
+      "format": "date-time"
+    },
+    "voucherNumber": {
+      "description": "Invoice number in FOLIO system",
```
[the remark this answers is not in the document]
Comment (craigmcnally) [reply mod-invoice-storage/schemas/batched_voucher.json]: Fixed in bathced_voucher as well as in voucher

File: mod-invoice-storage/schemas/batch_voucher.json [hunk 3]
```diff
@@ -0,0 +1,51 @@
+{
```
Comment (javalkar-varun) [inline mod-invoice-storage/schemas/batch_voucher.json]: If this is a collection then schema name should be `batch_voucher_collection.json` to be consistent with other schemas.

Comment (craigmcnally) [reply mod-invoice-storage/schemas/batch_voucher.json]: it's more than just a simple collection

File: mod-invoice-storage/schemas/voucher.json [hunk 4]
```diff
@@ -15,6 +15,12 @@
       "description": "The amount that is actually paid",
       "type": "number"
     },
+    "batchGroupId": {
+      "description": "UUID of the batch group to use when generating batch vouchers",
+      "type": "string",
+      "pattern": "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$",
+      "default": "2a2cb998-1437-41d1-88ad-01930aaeadd5"
+    },
     "batchNumber": {
```
Comment (javalkar-varun) [inline mod-invoice-storage/schemas/voucher.json]: As per [wiki](https://wiki.folio.org/pages/viewpage.action?spaceKey=FOLIJET&title=Batch+Voucher+Exports#BatchVoucherExports-invoice), should `batchNumber` field be removed as it is no longer required?

Comment (craigmcnally) [reply mod-invoice-storage/schemas/voucher.json]: Ah, thank you for the reminder... I just added that to the wiki last night after a discussion with Dennis

Comment (craigmcnally) [reply mod-invoice-storage/schemas/voucher.json]: I'll need to update the sample data and example as well.

**Comment by craigmcnally** [comment]
> It is necessary to update the schemes in the mod-finance, because the default value for a new field will be set by storage-module, if the mod-finance does not know about this field we will get an error.

Wait a second - for some reason the first time I read this I thought you were referring to mod-invoice, which makes sense - I created a PR for that.  But I just re-read this and see you actually say mod-finance - is that a typo?  I'm not sure I understand why mod-finance would need to be updated....  Does mod-finance ever deal with the invoice or voucher records?  I thought they only deal with references (UUIDs) to those records.

**Comment by pautau-s** [comment]
I got a little lost in our modules. It should definitely be a mod-invoice.

**Pull request merged**